File: go/orm/crud_test.go
package orm_test

import (
	"context"
	"errors"
	"testing"

	"github.com/kosmassp/SimpleOrm/go/orm"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/testsupport"
	"github.com/kosmassp/SimpleOrm/go/orm/sample"
)

func newTransaction(userID int64) *sample.Transaction {
	tx := &sample.Transaction{
		UserID: userID,
		Status: sample.Pending,
		Amount: orm.MustDecimal("10"),
	}
	tx.CreatedAtUtc = testsupport.SeedTime
	return tx
}

func TestUpdate_WritesTheFullRowByKey(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)
	ada := testsupport.InsertUser(t, db, "Ada", "ada@example.com")

	ada.Name = "Ada Lovelace"
	updated := testsupport.SeedTime
	ada.UpdatedAtUtc = &updated
	if err := orm.Update(ctx, db, ada); err != nil {
		t.Fatal(err)
	}

	loaded, err := orm.Get[sample.User](ctx, db, ada.ID)
	if err != nil {
		t.Fatal(err)
	}
	if loaded.Name != "Ada Lovelace" {
		t.Errorf("expected updated name, got %s", loaded.Name)
	}
	if loaded.UpdatedAtUtc == nil || !loaded.UpdatedAtUtc.Equal(testsupport.SeedTime) {
		t.Errorf("expected updated_at, got %v", loaded.UpdatedAtUtc)
	}
}

func TestUpdate_OfMissingRowIsCRUD001(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)
	ghost := &sample.User{ID: 999_999, Name: "Ghost", Email: "ghost@example.com"}
	ghost.CreatedAtUtc = testsupport.SeedTime

	err := orm.Update(ctx, db, ghost)
	if orm.CodeOf(err) != "CRUD-001" {
		t.Fatalf("expected CRUD-001, got %v", err)
	}
}

func TestUpdate_VersionedUpdateIncrementsAndDetectsConflicts(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)
	ada := testsupport.InsertUser(t, db, "Ada", "ada@example.com")
	if err := orm.Insert(ctx, db, newTransaction(ada.ID)); err != nil {
		t.Fatal(err)
	}

	first, err := orm.From[sample.Transaction](db).Where(orm.Eq("UserID", ada.ID)).Single(ctx)
	if err != nil {
		t.Fatal(err)
	}
	stale, err := orm.Get[sample.Transaction](ctx, db, first.ID)
	if err != nil {
		t.Fatal(err)
	}
	if first.Version != 0 {
		t.Fatalf("expected version 0, got %d", first.Version)
	}

	first.Status = sample.Completed
	if err := orm.Update(ctx, db, &first); err != nil {
		t.Fatal(err)
	}
	if first.Version != 1 {
		t.Fatalf("expected version bumped to 1 in memory, got %d", first.Version)
	}

	first.Amount = orm.MustDecimal("12")
	if err := orm.Update(ctx, db, &first); err != nil {
		t.Fatal(err)
	}
	if first.Version != 2 {
		t.Fatalf("expected version 2, got %d", first.Version)
	}

	stale.Amount = orm.MustDecimal("99") // still version 0
	err = orm.Update(ctx, db, &stale)
	if orm.CodeOf(err) != "CRUD-010" {
		t.Fatalf("expected CRUD-010, got %v", err)
	}

	current, err := orm.Get[sample.Transaction](ctx, db, first.ID)
	if err != nil {
		t.Fatal(err)
	}
	if !current.Amount.Equal(orm.MustDecimal("12")) {
		t.Errorf("expected the stale write to change nothing, got %s", current.Amount)
	}
	if current.Version != 2 {
		t.Errorf("expected version 2, got %d", current.Version)
	}
}

func TestDelete_ByKeyAndVersionedDeleteByEntity(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)
	ada := testsupport.InsertUser(t, db, "Ada", "ada@example.com")

	if err := orm.Delete[sample.User](ctx, db, ada.ID); err != nil {
		t.Fatal(err)
	}
	if missing, err := orm.GetOrDefault[sample.User](ctx, db, ada.ID); err != nil || missing != nil {
		t.Fatalf("expected nil, nil, got %v, %v", missing, err)
	}

	err := orm.Delete[sample.User](ctx, db, ada.ID)
	if orm.CodeOf(err) != "CRUD-001" {
		t.Fatalf("expected CRUD-001, got %v", err)
	}

	// Version-checked delete: a stale entity may not delete the row.
	grace := testsupport.InsertUser(t, db, "Grace", "grace@example.com")
	if err := orm.Insert(ctx, db, newTransaction(grace.ID)); err != nil {
		t.Fatal(err)
	}
	tx, err := orm.From[sample.Transaction](db).Where(orm.Eq("UserID", grace.ID)).Single(ctx)
	if err != nil {
		t.Fatal(err)
	}
	staleTx, err := orm.Get[sample.Transaction](ctx, db, tx.ID)
	if err != nil {
		t.Fatal(err)
	}

	tx.Status = sample.Cancelled
	if err := orm.Update(ctx, db, &tx); err != nil { // bumps the row to version 1
		t.Fatal(err)
	}

	err = orm.DeleteEntity(ctx, db, &staleTx)
	if orm.CodeOf(err) != "CRUD-010" {
		t.Fatalf("expected CRUD-010, got %v", err)
	}

	if err := orm.DeleteEntity(ctx, db, &tx); err != nil { // fresh version deletes fine
		t.Fatal(err)
	}
	if missing, err := orm.GetOrDefault[sample.Transaction](ctx, db, tx.ID); err != nil || missing != nil {
		t.Fatalf("expected nil, nil, got %v, %v", missing, err)
	}
}

func TestDelete_CompositeKeyByList(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)
	ada := testsupport.InsertUser(t, db, "Ada", "ada@example.com")
	role := &sample.Role{Name: "auditor"}
	role.CreatedAtUtc = testsupport.SeedTime
	if err := orm.Insert(ctx, db, role); err != nil {
		t.Fatal(err)
	}
	link := &sample.UserRole{UserID: ada.ID, RoleID: role.ID}
	link.CreatedAtUtc = testsupport.SeedTime
	if err := orm.Insert(ctx, db, link); err != nil {
		t.Fatal(err)
	}

	key := []any{ada.ID, role.ID}
	if err := orm.Delete[sample.UserRole](ctx, db, key); err != nil {
		t.Fatal(err)
	}
	if missing, err := orm.GetOrDefault[sample.UserRole](ctx, db, key); err != nil || missing != nil {
		t.Fatalf("expected nil, nil, got %v, %v", missing, err)
	}
}

func TestUpdateOnly_WritesTheListedColumnsAndNothingElse(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)
	ada := testsupport.InsertUser(t, db, "Ada", "ada@example.com")

	ada.Name = "Ada Lovelace"
	ada.Email = "changed@example.com" // set in memory, not listed
	if err := orm.UpdateOnly(ctx, db, ada, "Name"); err != nil {
		t.Fatal(err)
	}

	loaded, err := orm.Get[sample.User](ctx, db, ada.ID)
	if err != nil {
		t.Fatal(err)
	}
	if loaded.Name != "Ada Lovelace" {
		t.Errorf("expected the listed column written, got %s", loaded.Name)
	}
	if loaded.Email != "ada@example.com" {
		t.Errorf("expected the unlisted column untouched, got %s", loaded.Email)
	}
}

func TestUpdateOnly_KeepsRowLevelConcurrencyAndBumpsTheVersion(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)
	ada := testsupport.InsertUser(t, db, "Ada", "ada@example.com")
	if err := orm.Insert(ctx, db, newTransaction(ada.ID)); err != nil {
		t.Fatal(err)
	}
	first, err := orm.From[sample.Transaction](db).Where(orm.Eq("UserID", ada.ID)).Single(ctx)
	if err != nil {
		t.Fatal(err)
	}
	stale, err := orm.Get[sample.Transaction](ctx, db, first.ID)
	if err != nil {
		t.Fatal(err)
	}

	first.Amount = orm.MustDecimal("12")
	if err := orm.UpdateOnly(ctx, db, &first, "Amount"); err != nil {
		t.Fatal(err)
	}
	if first.Version != 1 {
		t.Fatalf("expected version bumped to 1, got %d", first.Version)
	}

	stale.Status = sample.Completed // a disjoint column, still version 0
	if err := orm.UpdateOnly(ctx, db, &stale, "Status"); orm.CodeOf(err) != "CRUD-010" {
		t.Fatalf("expected CRUD-010, got %v", err)
	}

	current, err := orm.Get[sample.Transaction](ctx, db, first.ID)
	if err != nil {
		t.Fatal(err)
	}
	if !current.Amount.Equal(orm.MustDecimal("12")) || current.Status != sample.Pending || current.Version != 1 {
		t.Errorf("stale write changed the row: %+v", current)
	}

	ghost := &sample.User{ID: 999_999, Name: "Ghost", Email: "ghost@example.com"}
	ghost.CreatedAtUtc = testsupport.SeedTime
	if err := orm.UpdateOnly(ctx, db, ghost, "Name"); orm.CodeOf(err) != "CRUD-001" {
		t.Fatalf("expected CRUD-001, got %v", err)
	}
}

func TestUpdateOnly_ValidatesTheListBeforeWriting(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)
	ada := testsupport.InsertUser(t, db, "Ada", "ada@example.com")
	if err := orm.Insert(ctx, db, newTransaction(ada.ID)); err != nil {
		t.Fatal(err)
	}
	tx, err := orm.From[sample.Transaction](db).Where(orm.Eq("UserID", ada.ID)).Single(ctx)
	if err != nil {
		t.Fatal(err)
	}

	cases := []struct {
		name       string
		call       func() error
		wantCode   string
		wantTarget string
	}{
		{"unknown", func() error { return orm.UpdateOnly(ctx, db, ada, "Nope") }, "CRUD-005", "User.Nope"},
		{"column name", func() error { return orm.UpdateOnly(ctx, db, ada, "name") }, "CRUD-005", "User.name"},
		{"key", func() error { return orm.UpdateOnly(ctx, db, ada, "ID") }, "CRUD-006", "User.ID"},
		{"version", func() error { return orm.UpdateOnly(ctx, db, &tx, "Version") }, "CRUD-006", "Transaction.Version"},
		{"empty", func() error { return orm.UpdateOnly(ctx, db, ada) }, "CRUD-007", "User"},
		{"repeat", func() error { return orm.UpdateOnly(ctx, db, ada, "Name", "Name") }, "CRUD-007", "User.Name"},
		{"read-only", func() error {
			return orm.UpdateOnly(ctx, db, &sample.UserTransactionTotal{UserName: "x"}, "UserName")
		}, "CRUD-003", "UserTransactionTotal"},
	}
	for _, tc := range cases {
		err := tc.call()
		if orm.CodeOf(err) != tc.wantCode {
			t.Errorf("%s: expected %s, got %v", tc.name, tc.wantCode, err)
			continue
		}
		var ormErr *orm.Error
		if errors.As(err, &ormErr) && ormErr.Target != tc.wantTarget {
			t.Errorf("%s: expected target %q, got %q", tc.name, tc.wantTarget, ormErr.Target)
		}
	}

	untouched, err := orm.Get[sample.User](ctx, db, ada.ID)
	if err != nil {
		t.Fatal(err)
	}
	if untouched.Name != "Ada" {
		t.Errorf("a refused list must write nothing, got %s", untouched.Name)
	}
}

func TestUpdate_GuardsReadOnlySources(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)

	err := orm.Update(ctx, db, &sample.UserTransactionTotal{UserName: "x"})
	if orm.CodeOf(err) != "CRUD-003" {
		t.Fatalf("expected CRUD-003, got %v", err)
	}

	err = orm.Delete[sample.UserTransactionTotal](ctx, db, int64(1))
	if orm.CodeOf(err) != "CRUD-003" {
		t.Fatalf("expected CRUD-003, got %v", err)
	}
}

// guidDoc is the client-GUID key strategy fixture (CrudTests.cs's GuidDoc).
type guidDoc struct {
	ID    orm.GUID `orm:"column,key"`
	Title string   `orm:"column"`
}

func (guidDoc) Entity() orm.EntityDef { return orm.EntityDef{Source: orm.Table("guid_docs")} }

func TestInsert_ClientGuidKeyStrategyRoundTrips(t *testing.T) {
	ctx := context.Background()
	db := testsupport.OpenSample(t)
	if err := orm.CreateTable[guidDoc](ctx, db); err != nil {
		t.Fatal(err)
	}

	doc := &guidDoc{Title: "spec"}
	if !doc.ID.IsZero() {
		t.Fatalf("expected a zero GUID before insert")
	}
	if err := orm.Insert(ctx, db, doc); err != nil {
		t.Fatal(err)
	}
	if doc.ID.IsZero() {
		t.Fatalf("expected a client-assigned GUID after insert")
	}

	loaded, err := orm.Get[guidDoc](ctx, db, doc.ID)
	if err != nil {
		t.Fatal(err)
	}
	if loaded.Title != "spec" {
		t.Errorf("expected spec, got %s", loaded.Title)
	}

	loaded.Title = "spec v2"
	if err := orm.Update(ctx, db, &loaded); err != nil {
		t.Fatal(err)
	}
	reloaded, err := orm.Get[guidDoc](ctx, db, doc.ID)
	if err != nil {
		t.Fatal(err)
	}
	if reloaded.Title != "spec v2" {
		t.Errorf("expected spec v2, got %s", reloaded.Title)
	}

	if err := orm.Delete[guidDoc](ctx, db, doc.ID); err != nil {
		t.Fatal(err)
	}
	if missing, err := orm.GetOrDefault[guidDoc](ctx, db, doc.ID); err != nil || missing != nil {
		t.Fatalf("expected nil, nil, got %v, %v", missing, err)
	}
}

File: php/tests/Session/CrudTest.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Session;

use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;
use SimpleOrm\Errors\ConcurrencyException;
use SimpleOrm\Errors\SimpleOrmException;
use SimpleOrm\Metadata\Attributes\Column;
use SimpleOrm\Metadata\Attributes\Generated;
use SimpleOrm\Metadata\Attributes\Key;
use SimpleOrm\Metadata\Attributes\Table;
use SimpleOrm\Metadata\ColumnType;
use SimpleOrm\Query\Criteria;
use SimpleOrm\Session\Db;
use SimpleOrm\Tests\Sample\Models\MonthlySalesTotal;
use SimpleOrm\Tests\Sample\Models\Role;
use SimpleOrm\Tests\Sample\Models\Transaction;
use SimpleOrm\Tests\Sample\Models\TransactionStatus;
use SimpleOrm\Tests\Sample\Models\User;
use SimpleOrm\Tests\Sample\Models\UserRole;
use SimpleOrm\Tests\Sample\Models\UserTransactionTotal;
use SimpleOrm\Tests\Support\SampleDatabase;
use SimpleOrm\Tests\Support\TempDatabase;
use SimpleOrm\Types\Decimal;

/**
 * Generated Update/Delete, optimistic concurrency, and the client-GUID key
 * strategy (§7.14-16). Mirrors dotnet's `CrudTests` + `SampleDomainTests`'
 * DDL/read-only-source scenarios.
 */
final class CrudTest extends TestCase
{
    private TempDatabase $fixture;

    private Db $db;

    protected function setUp(): void
    {
        $this->fixture = TempDatabase::create();
        $this->db = SampleDatabase::open($this->fixture);
    }

    protected function tearDown(): void
    {
        $this->db->close();
        $this->fixture->delete();
    }

    private static function newTransaction(int $userId): Transaction
    {
        $transaction = new Transaction();
        $transaction->userId = $userId;
        $transaction->status = TransactionStatus::Pending;
        $transaction->amount = Decimal::of('10');
        $transaction->createdAtUtc = SampleDatabase::seedTime();

        return $transaction;
    }

    #[Test]
    public function update_writes_the_full_row_by_key(): void
    {
        $ada = SampleDatabase::insertUser($this->db, 'Ada', 'ada@example.com');

        $ada->name = 'Ada Lovelace';
        $ada->updatedAtUtc = SampleDatabase::seedTime();
        $this->db->update($ada);

        $loaded = $this->db->get(User::class, $ada->id);
        self::assertSame('Ada Lovelace', $loaded->name);
        self::assertEquals(SampleDatabase::seedTime(), $loaded->updatedAtUtc);
    }

    #[Test]
    public function update_of_missing_row_is_crud001(): void
    {
        $ghost = new User();
        $ghost->id = 999_999;
        $ghost->name = 'Ghost';
        $ghost->email = 'ghost@example.com';
        $ghost->createdAtUtc = SampleDatabase::seedTime();

        try {
            $this->db->update($ghost);
            self::fail('expected CRUD-001');
        } catch (SimpleOrmException $e) {
            self::assertSame('CRUD-001', $e->errorCode);
        }
    }

    #[Test]
    public function versioned_update_increments_and_detects_conflicts(): void
    {
        $ada = SampleDatabase::insertUser($this->db, 'Ada', 'ada@example.com');
        $this->db->insert(self::newTransaction($ada->id));

        $first = $this->db->from(Transaction::class)->where(Criteria::eq('userId', $ada->id))->single();
        $stale = $this->db->get(Transaction::class, $first->id);
        self::assertSame(0, $first->version);

        $first->status = TransactionStatus::Completed;
        $this->db->update($first);
        self::assertSame(1, $first->version);   // bumped in memory (§7.16)

        $first->amount = Decimal::of('12');
        $this->db->update($first);   // sequential updates keep working
        self::assertSame(2, $first->version);

        $stale->amount = Decimal::of('99');   // still version 0
        try {
            $this->db->update($stale);
            self::fail('expected CRUD-010');
        } catch (ConcurrencyException $e) {
            self::assertSame('CRUD-010', $e->errorCode);
        }

        $current = $this->db->get(Transaction::class, $first->id);
        self::assertTrue($current->amount->equals(Decimal::of('12')));   // the stale write changed nothing
        self::assertSame(2, $current->version);
    }

    #[Test]
    public function update_only_writes_the_listed_columns_and_nothing_else(): void
    {
        $ada = SampleDatabase::insertUser($this->db, 'Ada', 'ada@example.com');

        $ada->name = 'Ada Lovelace';
        $ada->email = 'changed@example.com';   // set in memory, not listed
        $this->db->updateOnly($ada, ['name']);

        $loaded = $this->db->get(User::class, $ada->id);
        self::assertSame('Ada Lovelace', $loaded->name);
        self::assertSame('ada@example.com', $loaded->email);   // the unlisted column was not written
    }

    #[Test]
    public function update_only_keeps_row_level_concurrency_and_bumps_the_version(): void
    {
        $ada = SampleDatabase::insertUser($this->db, 'Ada', 'ada@example.com');
        $this->db->insert(self::newTransaction($ada->id));
        $first = $this->db->from(Transaction::class)->where(Criteria::eq('userId', $ada->id))->single();
        $stale = $this->db->get(Transaction::class, $first->id);

        $first->amount = Decimal::of('12');
        $this->db->updateOnly($first, ['amount']);
        self::assertSame(1, $first->version);

        $stale->status = TransactionStatus::Completed;   // a disjoint column, still version 0
        try {
            $this->db->updateOnly($stale, ['status']);
            self::fail('expected CRUD-010');
        } catch (ConcurrencyException $e) {
            self::assertSame('CRUD-010', $e->errorCode);
        }

        $current = $this->db->get(Transaction::class, $first->id);
        self::assertTrue($current->amount->equals(Decimal::of('12')));
        self::assertSame(TransactionStatus::Pending, $current->status);
        self::assertSame(1, $current->version);

        $ghost = new User();
        $ghost->id = 999_999;
        $ghost->name = 'Ghost';
        $ghost->email = 'ghost@example.com';
        $ghost->createdAtUtc = SampleDatabase::seedTime();
        try {
            $this->db->updateOnly($ghost, ['name']);
            self::fail('expected CRUD-001');
        } catch (SimpleOrmException $e) {
            self::assertSame('CRUD-001', $e->errorCode);
        }
    }

    #[Test]
    public function update_only_validates_the_list_before_writing(): void
    {
        $ada = SampleDatabase::insertUser($this->db, 'Ada', 'ada@example.com');
        $this->db->insert(self::newTransaction($ada->id));
        $tx = $this->db->from(Transaction::class)->where(Criteria::eq('userId', $ada->id))->single();

        $codeOf = function (object $entity, array $properties): string {
            try {
                $this->db->updateOnly($entity, $properties);
            } catch (SimpleOrmException $e) {
                return $e->errorCode;
            }

            return 'no error';
        };

        self::assertSame('CRUD-005', $codeOf($ada, ['nope']));
        self::assertSame('CRUD-005', $codeOf($ada, ['created_at']));   // property names, not column names
        self::assertSame('CRUD-006', $codeOf($ada, ['id']));
        self::assertSame('CRUD-006', $codeOf($tx, ['version']));
        self::assertSame('CRUD-007', $codeOf($ada, []));
        self::assertSame('CRUD-007', $codeOf($ada, ['name', 'name']));

        $readOnly = new UserTransactionTotal();
        self::assertSame('CRUD-003', $codeOf($readOnly, ['userName']));

        self::assertSame('Ada', $this->db->get(User::class, $ada->id)->name);
    }

    #[Test]
    public function delete_by_key_and_versioned_delete_by_entity(): void
    {
        $ada = SampleDatabase::insertUser($this->db, 'Ada', 'ada@example.com');

        $this->db->delete(User::class, $ada->id);
        self::assertNull($this->db->getOrDefault(User::class, $ada->id));

        try {
            $this->db->delete(User::class, $ada->id);
            self::fail('expected CRUD-001');
        } catch (SimpleOrmException $e) {
            self::assertSame('CRUD-001', $e->errorCode);
        }

        // Version-checked delete: a stale entity may not delete the row.
        $grace = SampleDatabase::insertUser($this->db, 'Grace', 'grace@example.com');
        $this->db->insert(self::newTransaction($grace->id));
        $tx = $this->db->from(Transaction::class)->where(Criteria::eq('userId', $grace->id))->single();
        $staleTx = $this->db->get(Transaction::class, $tx->id);

        $tx->status = TransactionStatus::Cancelled;
        $this->db->update($tx);   // bumps the row to version 1

        try {
            $this->db->delete(Transaction::class, $staleTx);
            self::fail('expected CRUD-010');
        } catch (ConcurrencyException $e) {
            self::assertSame('CRUD-010', $e->errorCode);
        }

        $this->db->delete(Transaction::class, $tx);   // fresh version deletes fine
        self::assertNull($this->db->getOrDefault(Transaction::class, $tx->id));
    }

    #[Test]
    public function composite_key_delete_by_tuple(): void
    {
        $ada = SampleDatabase::insertUser($this->db, 'Ada', 'ada@example.com');
        $role = new Role();
        $role->name = 'auditor';
        $role->createdAtUtc = SampleDatabase::seedTime();
        $this->db->insert($role);

        $link = new UserRole();
        $link->userId = $ada->id;
        $link->roleId = $role->id;
        $link->createdAtUtc = SampleDatabase::seedTime();
        $this->db->insert($link);

        $this->db->delete(UserRole::class, [$ada->id, $role->id]);
        self::assertNull($this->db->getOrDefault(UserRole::class, [$ada->id, $role->id]));
    }

    #[Test]
    public function writes_guard_readonly_sources(): void
    {
        $total = new UserTransactionTotal();

        try {
            $this->db->insert($total);
            self::fail('expected CRUD-003 (insert)');
        } catch (SimpleOrmException $e) {
            self::assertSame('CRUD-003', $e->errorCode);
        }

        try {
            $this->db->update($total);
            self::fail('expected CRUD-003 (update)');
        } catch (SimpleOrmException $e) {
            self::assertSame('CRUD-003', $e->errorCode);
        }

        try {
            $this->db->delete(UserTransactionTotal::class, 1);
            self::fail('expected CRUD-003 (delete)');
        } catch (SimpleOrmException $e) {
            self::assertSame('CRUD-003', $e->errorCode);
        }
    }

    #[Test]
    public function materialized_view_creation_on_sqlite_is_ddl002(): void
    {
        try {
            $this->db->createView(MonthlySalesTotal::class);
            self::fail('expected DDL-002');
        } catch (SimpleOrmException $e) {
            self::assertSame('DDL-002', $e->errorCode);
        }
    }

    #[Test]
    public function client_guid_key_strategy_round_trips(): void
    {
        $this->db->createTable(GuidDocFixture::class);

        $doc = new GuidDocFixture();
        $doc->title = 'spec';
        self::assertFalse(isset($doc->id));
        $this->db->insert($doc);
        self::assertNotSame('', $doc->id);   // client-assigned on insert

        $loaded = $this->db->get(GuidDocFixture::class, $doc->id);
        self::assertSame('spec', $loaded->title);

        $loaded->title = 'spec v2';
        $this->db->update($loaded);
        self::assertSame('spec v2', $this->db->get(GuidDocFixture::class, $doc->id)->title);

        $this->db->delete(GuidDocFixture::class, $doc->id);
        self::assertNull($this->db->getOrDefault(GuidDocFixture::class, $doc->id));
    }
}

/** @internal test fixture for the client-GUID key strategy — not a Sample model (CODING-STANDARD §7). */
#[Table('guid_docs')]
final class GuidDocFixture
{
    #[Key]
    #[Column(type: ColumnType::Guid)]
    public string $id;

    #[Column]
    public string $title;
}
